lib/TPP/Transforms/FoldIntoEltwise.cpp

```diff
@@ -105,10 +105,93 @@ struct BroadcastIntoEltwise
   }
 };
 
+// Folds linalg.max(linalg.fill(%cst), ...) into linalg.generic op.
+// For example, rewrites:
+//   %fill = linalg.fill %cst
+//   linalg.max %val %fill
+// into:
+//   linalg.generic
+//     arith.max %in, %cst
+//
+// NOTE: This could be achieved by controlled generalization + linalg eltwise
+//       operation fusion using the upstream linalg fusion pass and the upstream
+//       `linalg-inline-scalar-operands` pass extended to support scalar
+//       non-shaped inputs.
+// TODO: Linalg eltwise fusion pass could be generalized to operate on LinalgOp
+//       instead of GenericOp.
+// TODO: Extend linalg scalar inlining pass.
+struct FillIntoMax : public OpRewritePattern<linalg::MaxOp> {
+  using OpRewritePattern<linalg::MaxOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(linalg::MaxOp maxOp,
+                                PatternRewriter &rewriter) const override {
+    if (!maxOp.hasPureTensorSemantics())
+      return rewriter.notifyMatchFailure(maxOp, "expects tensor semantics");
+
+    // Look for fill within inputs.
+    if (llvm::none_of(maxOp.getInputs(), [](Value input) {
+          auto op = input.getDefiningOp();
+          return op && isa<linalg::FillOp>(op);
+        }))
+      return rewriter.notifyMatchFailure(maxOp, "no fill producers");
+
+    SmallVector<Value> inputs;
+    SmallVector<AffineMap> indexingMaps;
+    SmallVector<Value> constants;
+
+    for (auto [idx, input] : llvm::enumerate(maxOp.getInputs())) {
+      auto fillOp = input.getDefiningOp<linalg::FillOp>();
+      if (!fillOp) {
+        // Keep the operand as is.
+        inputs.push_back(input);
+        indexingMaps.push_back(
+            maxOp.getMatchingIndexingMap(maxOp.getDpsInputOperand(idx)));
+        continue;
+      }
+
+      // TODO: Check if there is any benefit in this over passing scalar
+      //       as an input argument.
+      // Store the constant separately to be later inserted directly into
+      // generic's body.
```

Comment:
For now I wanted to have the constant inlined but actually I'm not sure if it makes any real difference.
If it doesn't, this pass could be simplified and easy generalized.